### PR TITLE
feat(stealth): spend-time scripts for stealth UTXOs (TIP-0006)

### PR DESCRIPTION
**Pull request by sdbondi**
## Description

Implements [TIP-0006](https://github.com/tari-project/tips): a third `SpendCondition` variant, `Script`, that gates the spend of an Ootle stealth UTXO on a **stateless, sandboxed WASM predicate** rather than on the spender's credentials alone.

Unlike `Signed`/`AccessRule` (which answer *"is the spender authorized?"*), a spend script answers *"is this particular spend valid?"* by introspecting the spending `StealthTransferStatement` — its inputs, its outputs (including each output's own spend condition), revealed amounts and the current epoch. The predicate is an ordinary `is_mut == false` template function returning unit that **rejects by panicking**, modelled directly on the existing resource `AuthHook`. Covenants (timelocks, vaults, value constraints) are delivered as `template_lib` helpers over this — no separate covenant field.

## What's included

- **Types + ABI** — `SpendCondition::Script(SpendScript)` and `SpendContext` view types in `tari_template_lib_types`; `EngineOp::SpendContextInvoke` + `SpendContextAction`; `calc_stealth_statement_weight` now charges each output's serialized spend-condition bytes; regenerated TS bindings.
- **Read-only execution mode** — a `read_only` flag on `CallFrame` making `write_lock_substate` / `new_substate` return `WriteInReadOnlyContext`. The single write-lock chokepoint neutralises every mutating host op at once.
- **Engine** — `validate_spend_script_signature` enforced at **creation time (T1**, rejecting the creating tx so malformed conditions are un-creatable rather than permanently-locked funds**)** and **spend time (T2**, the authoritative, mandatory gate**)**; `RuntimeInterface::spend_context_invoke`; the predicate is invoked in a restricted read-only, no-cross-template frame with an empty auth scope.
- **template_lib** — `SpendContext` lazy accessors over the host op + covenant helpers (`require_output_preserves_condition`, `require_timelock`, `require_output_to`).
- **Wallet** — `PayTo::Script` builder support.
- **Tests + example** — a `spend_script` example template and engine tests covering the TIP test cases: timelock (allow/reject), recursive covenant (preserve/change/empty-outputs), signature lock (valid/invalid), read-only enforcement, sandbox deny-list, and creation-time validation (unknown/mutable/non-unit/missing-context/wrong-arg-count/malformed-CBOR).

## Security model

The sandbox has two layers, both enforced in the **core engine path** (not in optional runtime modules — security must not depend on which modules are registered):

- **(a) Write-lock chokepoint** — every state write funnels through `write_lock_substate` / `new_substate`, which refuse in a read-only frame. This is whitelist-grade: it neutralises every mutating host op (including future ones), so a predicate that directly calls e.g. `vault_invoke(Withdraw)` still fails.
- **(b) Deny-list for non-lock-mediated effects** — `call_invoke`, `generate_random_invoke`, `generate_uuid`, `emit_event`, `proof_invoke`, `bucket_invoke` are denied in the engine's core per-host-op hook (returning `ForbiddenInReadOnlyContext`). The list is restricted to WASM host ops (each `EngineOp`-backed); instruction-level ops like `pay_fee`/`publish_template` have no `EngineOp` and can never run in a read-only frame.

**T2 is mandatory**: a `SpendCondition` is untrusted substate data, so the function shape and bound-arg encoding are re-validated immediately before invoking, regardless of any earlier (T1) check. Together (a)+(b) make the predicate deterministic (no randomness; epoch fixed for the call) and provably side-effect-free.

## Versioning

Followed `scripts/crate_versioning.py impact tari_template_abi --breaking`: workspace bumped to `0.35.0` (tier-3 core crates) plus the template/wallet crates and their pins.

## Testing

`cargo check --workspace` passes. `nextest` sweeps on the affected crates (`tari_engine`, `tari_ootle_transaction`, `tari_template_lib_types`, wallet crates, `ootle-wasm-core`, `tari_transaction_manifest`, `tari_template_lib`) pass with no regressions, including the full existing `stealth.rs` suite and the new `spend_script.rs` tests.

### CODE REVIEW

File: crates/template_lib/src/spend_context.rs [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+    /// Recursive "stay in the vault" covenant: asserts that every stealth output carries the same spend condition that
+    /// invoked this predicate.
+    pub fn require_output_preserves_condition(&self) {
+        let me = self.invoking_condition();
+        assert!(
+            self.outputs().iter().all(|o| o.spend_condition == me),
+            "spend script covenant: every output must preserve the invoking spend condition",
+        );
+    }
```
[the remark this answers is not in the document]
Comment (sdbondi) [reply crates/template_lib/src/spend_context.rs:115]: Good catch — fixed in ca6c6489a. Added a `!outputs.is_empty()` guard so a fully-revealed spend with zero stealth outputs can no longer satisfy the covenant vacuously, plus a `covenant_rejects_spend_with_no_stealth_outputs` regression test. Also documented that the helper bounds output spend conditions only, not the revealed amount (compose with a revealed-output check for full value-locking).
